### PR TITLE
Ensure hotbackup agency lock is released under all circumstances.

### PR DESCRIPTION
**Pull request by neunhoef**
We have some reports that sometimes the hotbackup agency lock can be left behind. This was happening in https://arangodb.atlassian.net/browse/OASIS-252 and in other tickets. This PR addresses this problem although we never fully understood how this problem actually arises. In this PR, scope guards are used to ensure proper cleanup of the agency lock.

### CODE REVIEW

**Comment by neunhoef** [comment]
http://jenkins01.arangodb.biz:8080/view/PR/job/arangodb-matrix-pr/10644/

File: arangod/Cluster/ClusterMethods.cpp [hunk 1]
```diff
@@ -3750,10 +3750,17 @@ arangodb::Result hotBackupCoordinator(ClusterFeature& feature, VPackSlice const
       return result;
     }
 
+    auto releaseAgencyLock = scopeGuard([&]{
+      LOG_TOPIC("52416", INFO, Logger::BACKUP)
+          << "Releasing agency lock with scope guard! backupId: " << backupId;
```
Comment (jsteemann) [inline arangod/Cluster/ClusterMethods.cpp:3755]: I don't think the log level should be "INFO" here, rather DEBUG.

File: arangod/Cluster/ClusterMethods.cpp [hunk 2]
```diff
@@ -3766,6 +3773,7 @@ arangodb::Result hotBackupCoordinator(ClusterFeature& feature, VPackSlice const
 
     if (!result.ok()) {
       ci.agencyHotBackupUnlock(backupId, timeout, supervisionOff);
+      releaseAgencyLock.cancel();
```
Comment (jsteemann) [inline arangod/Cluster/ClusterMethods.cpp:3776]:
```suggestion
      // release the lock
      releaseAgencyLock.fire();
```

File: arangod/Cluster/ClusterMethods.cpp [hunk 3]
```diff
@@ -3750,10 +3750,17 @@ arangodb::Result hotBackupCoordinator(ClusterFeature& feature, VPackSlice const
...
+      LOG_TOPIC("52416", INFO, Logger::BACKUP)
+          << "Releasing agency lock with scope guard! backupId: " << backupId;
+      ci.agencyHotBackupUnlock(backupId, timeout, supervisionOff);
+    });
+
     if (end < steady_clock::now()) {
       LOG_TOPIC("352d6", INFO, Logger::BACKUP)
           << "hot backup didn't get to locking phase within " << timeout << "s.";
       auto hlRes = ci.agencyHotBackupUnlock(backupId, timeout, supervisionOff);
+      releaseAgencyLock.cancel();
```
Comment (jsteemann) [inline arangod/Cluster/ClusterMethods.cpp:3763]:
```suggestion
      // release the lock
      releaseAgencyLock.fire();
```

File: arangod/Cluster/ClusterMethods.cpp [hunk 4]
```diff
@@ -3785,6 +3793,7 @@ arangodb::Result hotBackupCoordinator(ClusterFeature& feature, VPackSlice const
         lockedServers.clear();
         if (result.is(TRI_ERROR_LOCAL_LOCK_FAILED)) {  // Unrecoverable
           ci.agencyHotBackupUnlock(backupId, timeout, supervisionOff);
+          releaseAgencyLock.cancel();
```
Comment (jsteemann) [inline arangod/Cluster/ClusterMethods.cpp:3796]:
```suggestion
          // release the lock
          releaseAgencyLock.fire();
```

File: arangod/Cluster/ClusterMethods.cpp [hunk 5]
```diff
@@ -3863,6 +3871,7 @@ arangodb::Result hotBackupCoordinator(ClusterFeature& feature, VPackSlice const
     if (!gotLocks && !allowInconsistent) {
       unlockDBServerTransactions(pool, backupId, dbServers);
       ci.agencyHotBackupUnlock(backupId, timeout, supervisionOff);
+      releaseAgencyLock.cancel();
```
Comment (jsteemann) [inline arangod/Cluster/ClusterMethods.cpp:3874]:
```suggestion
      // release the lock
      releaseAgencyLock.fire();
```

File: arangod/Cluster/ClusterMethods.cpp [hunk 6]
```diff
@@ -3881,6 +3890,7 @@ arangodb::Result hotBackupCoordinator(ClusterFeature& feature, VPackSlice const
     if (!result.ok()) {
       unlockDBServerTransactions(pool, backupId, dbServers);
       ci.agencyHotBackupUnlock(backupId, timeout, supervisionOff);
+      releaseAgencyLock.cancel();
```
Comment (jsteemann) [inline arangod/Cluster/ClusterMethods.cpp:3893]:
```suggestion
      // release the lock
      releaseAgencyLock.fire();
```

File: arangod/Cluster/ClusterMethods.cpp [hunk 7]
```diff
@@ -3892,17 +3902,17 @@ arangodb::Result hotBackupCoordinator(ClusterFeature& feature, VPackSlice const
 
     unlockDBServerTransactions(pool, backupId, dbServers);
     ci.agencyHotBackupUnlock(backupId, timeout, supervisionOff);
+    releaseAgencyLock.cancel();
```
Comment (jsteemann) [inline arangod/Cluster/ClusterMethods.cpp:3905]:
```suggestion
    // release the lock
    releaseAgencyLock.fire();
```

File: arangod/Cluster/ClusterMethods.cpp (hunk 1 above)
Comment (neunhoef) [reply arangod/Cluster/ClusterMethods.cpp]: Done

**Comment by neunhoef** [comment]
http://jenkins01.arangodb.biz:8080/view/PR/job/arangodb-matrix-pr/10742/

**Comment by jsteemann** [comment]
http://172.16.10.101:8080/view/PR/job/arangodb-matrix-pr/10808/

**Comment by neunhoef** [comment]
http://jenkins01.arangodb.biz:8080/view/PR/job/arangodb-matrix-pr/10949/

**Comment by neunhoef** [comment]
This is good to go.

**Pull request merged**